File: plugins/collector/event/apps/web/openresty.go
package web

import (
	"collector/cache/process"
	"collector/event/apps"
	"strings"
)

type Openresty struct {
	version string
}

func (Openresty) Name() string { return "openresty" }

func (Openresty) Type() string { return "web" }

func (n Openresty) Version() string { return n.version }

func (Openresty) Match(p *process.Process) bool {
	// As default, nginx runs with master_process
	// ignore the other processes, and report the worker process if we need
	return p.Name == "openresty"
}

// Tengine matches too
func (n *Openresty) Run(p *process.Process) (m map[string]string, err error) {
	result, err := apps.Execute(p, "-v")
	if err != nil {
		return
	}
	n.version = strings.TrimPrefix(result, "openresty/")
	return
}

func init() { apps.Regist(&Openresty{}) }

File: plugins/collector/src/event/apps/web/openresty.rs
use crate::event::apps::{execute, AppProc, IApp};
use std::collections::HashMap;

pub struct Openresty {
    version: String,
}

impl Openresty {
    pub fn new() -> Self { Openresty { version: String::new() } }
}

impl IApp for Openresty {
    fn name(&self)     -> &'static str { "openresty" }
    fn app_type(&self) -> &'static str { "web" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "openresty" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-v"])?;
        // Output may be on stdout or stderr, may be prefixed:
        //   "openresty/1.21.4.1"  or  "openresty version openresty/1.21.4.1"
        if let Some(pos) = out.find("openresty/") {
            let rest = &out[pos + "openresty/".len()..];
            self.version = rest.split(|c: char| c.is_whitespace()).next().unwrap_or("").to_owned();
        }
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
